### PR TITLE
feat(j-s): Deliver Subpoena to Court

### PR DESCRIPTION
**Pull request by gudjong**
# Deliver Subpoena to Court

[Skila staðfestu fyrirkalli í Auði](https://app.asana.com/0/1199153462262248/1208360072306898/f)

## What

- When the court issues a subpoena it is now delivered to the district court.

## Why

- The court wants to get the confirmed subpoean.

## Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] Formatting passes locally with my changes
- [x] I have rebased against main before asking for a review

## Release Notes

- **New Features**
  - Added support for delivering subpoenas to court with a new endpoint.
  - Introduced a new message type for handling court subpoenas.

- **Enhancements**
  - Improved message handling for new subpoenas, generating separate notifications for police and court.
  - Enhanced validation logic for case handling by implementing additional guards.
  - Integrated court document handling within the subpoena service.

- **Bug Fixes**
  - Refined error handling in various methods related to case updates and subpoena deliveries.

- **Tests**
  - Expanded test coverage for case updates and subpoena deliveries, ensuring robust validation of guard configurations and error scenarios.

### CODE REVIEW

File: apps/judicial-system/backend/src/app/modules/case/case.service.ts [hunk 1]
```diff
@@ -1369,7 +1380,10 @@ export class CaseService {
           await this.addMessagesForCourtCaseConnectionToQueue(updatedCase, user)
         }
       } else {
-        if (updatedCase.prosecutorId !== theCase.prosecutorId) {
+        if (
+          !isIndictment &&
```
Comment (gudjong) [inline apps/judicial-system/backend/src/app/modules/case/case.service.ts:1384]: This is not part of the current task. I just noticed this a fixed it. The prosecutor is deliverd to the district court in a different way for indictment cases.

File: apps/judicial-system/backend/src/app/modules/case/internalCase.controller.ts [hunk 2]
```diff
@@ -116,7 +116,10 @@ export class InternalCaseController {
     )
   }
 
-  @UseGuards(CaseExistsGuard)
+  @UseGuards(
+    CaseExistsGuard,
+    new CaseTypeGuard([...restrictionCases, ...investigationCases]),
+  )
```
Comment (gudjong) [inline apps/judicial-system/backend/src/app/modules/case/internalCase.controller.ts:122]: This is not part of the current task. I just noticed this a fixed it. The prosecutor is deliverd to the district court in a different way for indictment cases.

File: apps/judicial-system/backend/src/app/modules/subpoena/subpoena.service.ts [hunk 3]
```diff
@@ -356,6 +358,41 @@ export class SubpoenaService {
...
+      .catch((reason) => {
+        // Tolerate failure, but log error
+        this.logger.warn(
+          `Failed to upload subpoena ${subpoena.id} pdf to court for defendant ${defendant.id} of case ${theCase.id}`,
+          { reason },
+        )
+
+        return { delivered: false }
+      })
+  }
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/subpoena/subpoena.service.ts:394]: I like my way

File: apps/judicial-system/backend/src/app/modules/case/internalCase.controller.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/internalCase.controller.ts:122]: Good idea, but I am not going to change this now..

File: apps/judicial-system/backend/src/app/modules/case/test/internalCaseController/deliverProsecutorToCourtGuards.spec.ts [hunk 4]
```diff
@@ -14,19 +18,12 @@ describe('InternalCaseController - Deliver prosecutor to court guards', () => {
...
-    })
-
-    it('should have CaseExistsGuard as guard 1', () => {
-      expect(guard).toBeInstanceOf(CaseExistsGuard)
+  it('should have the right guard configuration', () => {
+    expect(guards).toHaveLength(2)
+    expect(new guards[0]()).toBeInstanceOf(CaseExistsGuard)
+    expect(guards[1]).toBeInstanceOf(CaseTypeGuard)
+    expect(guards[1]).toEqual({
+      allowedCaseTypes: [...restrictionCases, ...investigationCases],
```
[the remark this answers is not in the document]
Comment (gudjong) [reply apps/judicial-system/backend/src/app/modules/case/test/internalCaseController/deliverProsecutorToCourtGuards.spec.ts:26]: Good idea, but I am not changing this now.

File: apps/judicial-system/backend/src/app/modules/subpoena/test/internalSubpoenaController/deliverSubpoenaToPolice.spec.ts [hunk 5]
```diff
@@ -22,57 +22,59 @@ describe('InternalSubpoenaController - Deliver subpoena to police', () => {
...
 
     it('should call deliverSubpoenaToPolice', () => {
-      expect(mockSubpoenaService.deliverSubpoenaToPolice).toHaveBeenCalledWith(
+      expect(mockPdfService.getSubpoenaPdf).toBeCalledWith(
         theCase,
         defendant,
         subpoena,
-        user.user,
       )
+      // TODO: complege tests when all indictments are generated
```
Comment (unakb) [inline apps/judicial-system/backend/src/app/modules/subpoena/test/internalSubpoenaController/deliverSubpoenaToPolice.spec.ts:77]:
```suggestion
      // TODO: complete tests when all indictments are generated
```

**Review by unakb — APPROVED** [review]

**Pull request merged**